Libplanet.Explorer/Indexing/IBlockChainIndex.cs

```diff
@@ -0,0 +1,228 @@
+    /// containing <see cref="Block{T}"/>.</param>
+    /// <param name="containedBlock">The <see cref="BlockHash"/> of the indexed block that contains
+    /// the <see cref="Transaction{T}"/> with the <paramref name="txId"/>, if it exists.</param>
+    /// <returns>Whether the retrieval was successful.</returns>
+    bool TryGetContainedBlockHashById(TxId txId, out BlockHash containedBlock);
+
+    /// <summary>
+    /// Attempt to get the <see cref="BlockHash"/> of the indexed block that contains the
+    /// <see cref="Transaction{T}"/> with the <paramref name="txId"/>, and return the retrieved
+    /// <see cref="BlockHash"/> if it exists, or <c>null</c> if it does not.
+    /// </summary>
+    /// <param name="txId">The <see cref="TxId"/> of the <see cref="Transaction{T}"/> to find the
+    /// containing <see cref="Block{T}"/>.</param>
+    /// <returns>The <see cref="BlockHash"/> of the block containing the
+    /// <see cref="Transaction{T}"/> with the <paramref name="txId"/> if it exists, or <c>null</c>
+    /// if it does not.</returns>
+    Task<BlockHash?> TryGetContainedBlockHashByIdAsync(TxId txId);
+
+    /// <summary>
+    /// Get the last tx nonce of the <paramref name="address"/> that was indexed.
+    /// </summary>
+    /// <param name="address">The address to retrieve the tx nonce.</param>
+    /// <returns>The last tx nonce of the <paramref name="address"/> that was indexed.</returns>
+    /// <remarks>This method does not return the tx nonces of <see cref="Transaction{T}"/>s that
+    /// are currently staged.</remarks>
+    long? GetLastNonceByAddress(Address address);
+
+    /// <inheritdoc cref="GetLastNonceByAddress"/>
+    Task<long?> GetLastNonceByAddressAsync(Address address);
+
+    /// <summary>
+    /// Index the metadata of a <see cref="Block{T}"/> corresponding to the given
+    /// <paramref name="blockDigest"/>.
+    /// </summary>
+    /// <param name="blockDigest">The block digest object to index.</param>
+    /// <param name="txs">An <see cref="IEnumerable{T}"/> containing the <see cref="ITransaction"/>
+    /// instances corresponding to <see cref="BlockDigest.TxIds"/> of given
+    /// <paramref name="blockDigest"/>.</param>
+    /// <param name="token">A token to mark the cancellation of processing.</param>
+    /// <exception cref="IndexMismatchException">Thrown if the index already has seen a block in
+    /// the height of the given block, but the hash of the indexed block and the given block is
+    /// different.</exception>
+    internal void Index(
+        BlockDigest blockDigest, IEnumerable<ITransaction> txs, CancellationToken token);
+
+    /// <inheritdoc cref="Index"/>
+    internal Task IndexAsync(
+        BlockDigest blockDigest, IEnumerable<ITransaction> txs, CancellationToken token);
+
+    internal void SynchronizeForever<T>(BlockChain<T> chain, CancellationToken stoppingToken)
+        where T : IAction, new();
```

Comment:
`SynchronizeForever<T>()` doesn't seem to be used atm.

Reply:
I created it for the sake of symmetry, but do you reckon that it's unnecessary?

Reply:
Yes 😅